nixos/modules/services/networking/wireguard.nix

```diff
@@ -419,6 +483,116 @@ let
+      {
+        enable = values.peersAnnouncing.enable;
+        listenStreams = [(toString values.peersAnnouncing.port)];
+        socketConfig.Accept = true;
+        socketConfig.BindToDevice = name;
+        socketConfig.IPAddressAllow = map (v: v.allowedIPs) values.peers;
+        socketConfig.IPAddressDeny = "any";
+        socketConfig.MaxConnectionsPerSource = 1;
+        socketConfig.ReusePort = true;
+        wantedBy = [ "sockets.target" ];
+      };
+
+  generatePeersAnnouncingUnit = name: values:
+    nameValuePair "wireguard-${name}-peers-announcing@"
+      {
+        description = "WireGuard Peers Announcing - ${name}";
+        requires = [ "wireguard-${name}.service" ];
+        after = [ "wireguard-${name}.service" ];
+        path = with pkgs; [ wireguard-tools ];
+
+        serviceConfig = mkMerge [
+          peerUpdateSecurity
+          {
+            Type = "simple";
+            StandardInput = "null";
+            StandardOutput = "socket";
+            RestrictAddressFamilies = "";
+          }
+          (mkIf (values.interfaceNamespace != null)
+            { NetworkNamespacePath = "/var/run/netns/${values.interfaceNamespace}"; })
+        ];
+
+        # Send each configured peer, keeping only the fields: public_key and endpoint.
+        script = "wg show '${name}' dump | tail -n +2 | cut -f1,3";
+      };
+
+  generateEndpointReceivingUnit = { interfaceName, interfaceCfg, peer }: let
+      dynamicRefreshEnabled = peer.dynamicEndpointRefreshSeconds != 0;
+      peerService = peerUnitServiceName interfaceName peer.publicKey dynamicRefreshEnabled;
+    in
+    nameValuePair "wireguard-${interfaceName}-endpoint-receiving-${keyToUnitName peer.publicKey}"
+      {
+        description = "WireGuard ${interfaceName} Endpoint Receiving from ${peer.publicKey}";
+        requires = [ "${peerService}.service" ];
+        after = [ "${peerService}.service" ];
+        wantedBy = [ "${peerService}.service" ];
+        path = with pkgs; [ netcat-gnu wireguard-tools ];
+
+        serviceConfig = mkMerge [
+          peerUpdateSecurity
+          {
+            Type = "simple";
+            IPAddressAllow = [ peer.endpointReceiving.addr ];
+            RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_NETLINK" ];
+            Restart = "on-failure";
+          }
+          (mkIf (interfaceCfg.interfaceNamespace != null)
+            { NetworkNamespacePath = "/var/run/netns/${interfaceCfg.interfaceNamespace}"; })
+        ];
+
+        # Query the peer announcing other peers,
+        # for setting the current endpoint of all configured peers.
+        script = ''
+          declare -A configured_keys
+          configured_keys=(${concatMapStringsSep " " (p:
+            optionalString (p.publicKey != peer.publicKey) "[${p.publicKey}]=1")
+            interfaceCfg.peers})
+          while true; do
+            exec < <(exec nc </dev/null --tcp --wait=${toString peer.endpointReceiving.refreshSeconds} \
```

Comment:
What does `exec < ...)` do?

Reply:
Replaces stdin with the given file for the rest of the script (or until the next `exec <`). In this case, it's replaced with a process substitution, i.e. the output of the `nc` command is piped into the remainder of the script.

Reply:
Ah, so you can write

```sh
exec < ...;
while read ...
  # more script here
done
```

where `exec` is a standalone statement without subsequent command; from `man bash`, `If command  is not specified, any  redirections  take effect in the current shell`.

I wasn't aware of that, only used it as `execve()` wrapper so far.

Reply:
Just to mention that I've copied this pattern from [wg-json](https://github.com/WireGuard/wireguard-tools/blob/master/contrib/json/wg-json).